### PR TITLE
[SPARK-34726][SQL][2.4] Fix collectToPython timeouts

### PR DESCRIPTION
**Pull request by peter-toth**
### What changes were proposed in this pull request?

One of our customers frequently encounters `"serve-DataFrame" java.net.SocketTimeoutException: Accept timed` errors in PySpark because `DataSet.collectToPython()` in Spark 2.4 does the following:
1. Collects the results
2. Opens up a socket server that is then listening to the connection from Python side
3. Runs the event listeners as part of `withAction` on the same thread as SPARK-25680 is not available in Spark 2.4
4. Returns the address of the socket server to Python
5. The Python side connects to the socket server and fetches the data

As the customer has a custom, long running event listener the time between 2. and 5. is frequently longer than the default connection timeout and increasing the connect timeout is not a good solution as we don't know how long running the listeners can take.

### Why are the changes needed?

This PR simply moves the socket server creation (2.) after running the listeners (3.). I think this approach has has a minor side effect that errors in socket server creation are not reported as `onFailure` events, but currently errors happening during opening the connection from Python side or data transfer from JVM to Python are also not reported as events so IMO this is not a big change.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Added new UT + manual test.

### CODE REVIEW

**Comment by peter-toth** [comment]
cc @HyukjinKwon

**Comment by peter-toth** [comment]
cc @gaborgsomogyi

**Comment by peter-toth** [comment]
cc @attilapiros

**Comment by gaborgsomogyi** [comment]
Hmmm, this is exactly one reason why we've added this: https://github.com/apache/spark/pull/30389
Good to see that this has been solved in 3.x line with a separate thread but it would be good to solve it on 2.x line too.

**Comment by gaborgsomogyi** [comment]
Since I've seen a thread about 2.4.8 it would be good to solve this there, right @HyukjinKwon @dongjoon-hyun ?
I'm asking it whether we should go further with the in-depth review or not.

**Review by HyukjinKwon — COMMENTED** [review]
Okay, the fix looks making sense to me.

**Comment by HyukjinKwon** [comment]
cc @viirya FYI

**Comment by gaborgsomogyi** [comment]
I would like to torture this a bit but I agree with the main intention too.

**Comment by gaborgsomogyi** [comment]
`KafkaContinuousSinkSuite` failure is unrelated.

**Review by gaborgsomogyi — APPROVED** [review]
LGTM. Validated manually and passed, additionally don't see high risk in it.

**Comment by srowen** [comment]
Jenkins test this please

**Review by attilapiros — COMMENTED** [review]
The code looks good. But I think you can come up with a good unit test where a longer running `onSuccess` method is given with the listener (longer than a socket timeout...). 

WDYT?

**Comment by attilapiros** [comment]
> You mean measuring time as assert criteria?

No. Just having a `onSuccess` method with a longer sleep than socket timeout itself. 
We expect the data will be still available at the python side or even we should not leave the JVM and the test can be in one of the existing scala Suites. 

If this turn out to be too complex I can let it go. But let's check it first.

**Comment by peter-toth** [comment]
> > You mean measuring time as assert criteria?
> 
> No. Just having a `onSuccess` method with a longer sleep than socket timeout itself.
> We expect the data will be still available at the python side or even we should not leave the JVM and the test can be in one of the existing scala Suites.
> 
> If this turn out to be too complex I can let it go. But let's check it first.

Thanks, added in https://github.com/apache/spark/pull/31818/commits/6b18cc752379387dd13c009490b54be823cf2437

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala
Comment (gaborgsomogyi) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala:1616]: I would put this into finally to be on the safe side.

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala
Comment (attilapiros) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala:1600]: Let's extract the 15000 into a `private[spark] var` as member:

https://github.com/apache/spark/blob/6b18cc752379387dd13c009490b54be823cf2437/core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala#L899

And add a comment  above `// visible for testing`.
Then we can speed up this single test.

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala
Comment (gaborgsomogyi) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala:1601]: Just for my own understanding does this mean the test waits 15 seconds to pass?

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala
Comment (peter-toth) [reply sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala]: All right, added in https://github.com/apache/spark/pull/31818/commits/8f6b811a28d1813d72151fe9063d5794b99fa481

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala
Comment (peter-toth) [reply sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala]: Yes, we need to wait a bit longer than the timeout. Without the fix in `Dataset.collectToPython` this UT fails.

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala
Comment (peter-toth) [reply sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala]: Ok, added in https://github.com/apache/spark/commit/6b18cc752379387dd13c009490b54be823cf2437 and https://github.com/apache/spark/pull/31818/commits/8f6b811a28d1813d72151fe9063d5794b99fa481

**Comment by gaborgsomogyi** [comment]
Let's see jenkins...

**Comment by dongjoon-hyun** [comment]
Thank you for pinging me, @gaborgsomogyi . Yes, it's a good timing to discuss this and the decision is up to the release manager, @viirya .
> Since I've seen a thread about 2.4.8 it would be good to solve this there, right @HyukjinKwon @dongjoon-hyun ?
> I'm asking it whether we should go further with the in-depth review or not.

BTW, we should be more careful because 2.4.8 is the last release.

**Review by attilapiros — APPROVED** [review]
LGTM

**Comment by viirya** [comment]
> This PR simply moves the socket server creation (2.) after running the listeners (3.). I think this approach has has a minor side effect that errors in socket server creation are not reported as onFailure events, but currently errors happening during opening the connection from Python side or data transfer from JVM to Python are also not reported as events so IMO this is not a big change.

Hmm, I'm not sure about this. It sounds like if any people rely `onFailure` events, there will be a breaking change for them. As 2.4.8 is the last release and is only maintenance release, above seems to be a behavior change so it is still risky to me.

**Comment by peter-toth** [comment]
> > This PR simply moves the socket server creation (2.) after running the listeners (3.). I think this approach has has a minor side effect that errors in socket server creation are not reported as onFailure events, but currently errors happening during opening the connection from Python side or data transfer from JVM to Python are also not reported as events so IMO this is not a big change.
> 
> Hmm, I'm not sure about this. It sounds like if any people rely `onFailure` events, there will be a breaking change for them. As 2.4.8 is the last release and is only maintenance release, above seems to be a behavior change so it is still risky to me.

I think it is very rare that a `SocketServer` throws an exception because it can't bind to some port on `127.0.0.1`. Other communication errors between Java and Python are also not reported currently...

**Comment by viirya** [comment]
> I think it is very rare that a `SocketServer` throws an exception because it can't bind to some port on `127.0.0.1`. Other communication errors between Java and Python are also not reported as `onFailure` event currently...

Oh, I see. If only socket server creation is not caught by the listener, seems less risky. Okay it looks making sense. Would like to have more eyes on this. cc @cloud-fan @maropu

**Comment by gaborgsomogyi** [comment]
@viirya you see it well and you're right, it's a tradeoff which should be discussed in-depth to come up with balanced decision.
Pro:
* Used don't need to guess how much time listeners will consume and mustn't make guesses to set `spark.python.authenticate.socketTimeout` properly.

Con:
* If `ServerSocket` can't bind to localhost then an event will be dropped which is behavior change.
* `2.4.8` is the last `2.4` release so sudden movements are not suggested.

Couple of heavy users complained about this but that said `spark.python.authenticate.socketTimeout` can be increased as workaround. If we decide not to merge this then we must mention in our doc that such cases the timeout must be increased to a best guess timeout (this is only true for `2.4` line, because on `3.x` it runs in separate thread).
I've seen cases where this timeout was between 20 to 60 seconds and the complain was exactly this wide range.
I see their point because configuring huge and sensitive computations in PROD based on best guesses freaks out the guys.

**Comment by attilapiros** [comment]
> * Used don't need to guess how much time listeners will consume and mustn't make guesses to set `spark.python.authenticate.socketTimeout` properly.

Moreover `spark.python.authenticate.socketTimeout` is only available from 3.1.0 so they even cannot use that config as a workaround here.

**Comment by gaborgsomogyi** [comment]
True, so backport needed such case. I guess it wouldn't be brain surgery though.

**Comment by peter-toth** [comment]
@cloud-fan, @maropu could you please review this PR?

**Review by HyukjinKwon — APPROVED** [review]
LGTM. I'll merge in few days if there are no more comments.

**Review by viirya — APPROVED** [review]
I'm fine with this fix if no one raises objection.

**Comment by viirya** [comment]
Thanks all. Merging to 2.4.

**Comment by peter-toth** [comment]
Thanks all for the review.

**Comment by viirya** [comment]
BTW, the JIRA description is empty. I just copied the description here to the JIRA.

**Pull request closed without merging**
